staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -912,8 +912,6 @@ func (c *Cacher) dispatchEvents() {
 			// Never send a bookmark event if we did not see an event here, this is fine
 			// because we don't provide any guarantees on sending bookmarks.
 			if lastProcessedResourceVersion == 0 {
-				// pop expired watchers in case there has been no update
-				c.bookmarkWatchers.popExpiredWatchers()
```

Comment:
I see this as potentially problematic.

Imagine a usecase where no object is actually changing for a long time, but we potentially have a lot of watchers registered.
[FWIW - secrets are a good example]

After this change, until we get the first event, we will be accumulating all watchers in `bookmarkWatchers`.
And those currently are not deleted from `bookmarkWatcher` on watches being closed - they are just are cleaned up here and on dispatching a regular events.

So in this case, we will be accumulating stuff in `bookmarkWatchers`, in a pathological case leading to OOM.

We can't just delete this line - we need a mechanism to prevent it.

Reply:
Oh, I didn't realise that watchers won't be removed from `bookmarkWatcher` when they are closed. 
I think I would change the code to just "pop" closed watchers, would you agree ?

I think that memory footprint will not be huge. Since we are storing pointers to the `cacheWatcher` struct we should be able to store ~8K entires in just 1MB (assuming the size of the struct is ~100 bytes).

Given the above calculation I think we should be fine.
I could provide a new method `c.bookmarkWatchers.popClosedWatchers()` and just call it in line 916.

Reply:
> I think I would change the code to just "pop" closed watchers, would you agree ?

I think we need to ensure that we won't be iterating over too many (still active) watchers every second.
So I think it's not as simple as that.

I think that the "time=0" is somewhat special here...
1) For time>0, probably calling `popExpiredWatchers` as it was before is still the right thing to do?
2) If we do that, we're probably left only with those with `time=0`. For these closing stopped seems good enough, but we need to somehow ensure that we won't be iterating over all of the watchers on every loop...

Reply:
Actually, I've just realised that the cache receives periodic updates from the `etcd` every X seconds.
These periodic updates are reflected in `lastProcessedResourceVersion`.

Since the graduation of `EfficientWatchResumption` every Kube installation supports progress notification from etcd. In addition to that I think that the default notification period is `5 seconds`. 

So, that means we will be accumulating disconnected watcher for `5 s` in the worst case.
Expired watchers are not re-added again to `watchersBuffer`.

I think we should be fine with what we have atm, wdyt?

Reply:
Hmm - under the assumption that etcd is configured to really deliver them.
I can imagine a situation of some operator configuring those progress notifiy events to be delivered every 1h or sth like that. in which case it may become problematic...

I think that even we assume that those with time=0 aren't problematic, I would feel more comfortable if we keep calling `popExpiredWatchers` for time>0...

Reply:
OK, updated the PR as we discussed offline.